File: desktop/lib/dto/paginated_response.dart
class PaginatedResponse<T> {
  List<T> data;
  int pageIndex;
  int totalPages;
  bool hasPreviousPage;
  bool hasNextPage;

  PaginatedResponse({
    required this.data,
    required this.pageIndex,
    required this.totalPages,
    required this.hasPreviousPage,
    required this.hasNextPage,
  });

  factory PaginatedResponse.fromJson(
      Map<String, dynamic> json, T Function(Map<String, dynamic>) fromJsonT) {
    return PaginatedResponse(
      data: List<T>.from(json["data"].map((x) => fromJsonT(x))),
      pageIndex: json["pageIndex"],
      totalPages: json["totalPages"],
      hasPreviousPage: json["hasPreviousPage"],
      hasNextPage: json["hasNextPage"],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "data": List<dynamic>.from(data.map((x) => (x as dynamic).toJson())),
      "pageIndex": pageIndex,
      "totalPages": totalPages,
      "hasPreviousPage": hasPreviousPage,
      "hasNextPage": hasNextPage,
    };
  }
}

File: backend/api/DTO/PaginatedResponse.cs
namespace api.DTO;
public class PaginatedResponse<T>
{
    public List<T> Data { get; }
    public int PageIndex { get; }
    public int TotalPages { get; }
    public bool HasPreviousPage => PageIndex > 1;
    public bool HasNextPage => PageIndex < TotalPages;

    public PaginatedResponse(List<T> items, int pageIndex, int totalPages)
    {
        Data = items;
        PageIndex = pageIndex;
        TotalPages = totalPages;
    }
}